File: src/main/java/com/codingbaby/ohmyidea/key/CommandHolder.java
package com.codingbaby.ohmyidea.key;

import javax.swing.*;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Created by baby on 15/6/24.
 */

public class CommandHolder {

    private Map<String, CommandMapping> data = new LinkedHashMap<String, CommandMapping>();


    public void add(String key, String action, String desc) {
        data.put(key, new CommandMapping(key, new CommandNode(action), desc));
    }

    public CommandNode get(KeyStroke stroke) {
        char keyChar = stroke.getKeyChar();
        CommandMapping commandMapping = data.get("" + keyChar);
        if (commandMapping == null) {
            return null;
        }
        return commandMapping.getCommandNode();
    }

    public CommandNode get(String key) {
        CommandMapping commandMapping = data.get(key);
        if (commandMapping == null) {
            return null;
        }
        return commandMapping.getCommandNode();
    }

    public Map<String,String> getCommandDesc() {
        Map<String,String> map = new HashMap<String, String>();
        for (CommandMapping commandMapping : data.values()) {
            map.put(commandMapping.getKey(), commandMapping.getDesc());
        }
        return map;
    }
}

File: src/main/java/com/codingbaby/ohmyidea/key/CommandHolder.kt
package com.codingbaby.ohmyidea.key

import javax.swing.*
import java.util.HashMap
import java.util.LinkedHashMap

class CommandHolder {

    class CommandMapping(var key: String, val commandNode: CommandNode, val desc: String)


    private val data = LinkedHashMap<String, CommandMapping>()

    fun clear() {
        data.clear()
    }

    fun add(key: String, action: String, desc: String) {
        data.put(key, CommandMapping(key, CommandNode(action), desc))
    }

    operator fun get(stroke: KeyStroke): CommandNode? {
        val keyChar = stroke.keyChar
        val commandMapping = data["" + keyChar] ?: return null
        return commandMapping.commandNode
    }

    operator fun get(key: String): CommandNode? {
        val commandMapping = data[key] ?: return null
        return commandMapping.commandNode
    }

    val commandDesc: Map<String, String>
        get() {
            val map = HashMap<String, String>()
            for (commandMapping in data.values) {
                map.put(commandMapping.key, commandMapping.desc)
            }
            return map
        }
}
